### PR TITLE
Extract cluster sina function

### PR DESCRIPTION
**Pull request by russHyde**
This decomposes part of the `treeview()` function.

There was originally an inner-function `.pl_cluster_sina()` inside `treeview()`. This generated files that were stored in `<output_dir>/sina-<summarised_variable>.html` (for example, "sina-logistic_growth_rate.html") and some ancillary files required to display the contained htmlwidget.

State-dependent/modifying parts of .pl_cluster_sina were a call to "htmlwidgets::saveWidget" and a call to rnorm().

Variables pulled from the outer scope by .pl_cluster_sina were "output_dir".

A new function (`plot_cluster_sina`) was added to replace `.pl_cluster_sina`, the new function takes the same arguments as `.pl_cluster_sina` but also accepts `output_dir` argument.

Three main phases were identified and a separate function was added for each:
- conversion of the tree-plot dataframe (pldf) into a data-frame that could be used in the sina/cluster scatter plots
- creation of an interactive ggplot2 object that contained the scatter plot
- conversion of that scatter plot into an htmlwidget

The final phase of the original function saved the htmlwidget to a file (the htmlwidget/plot/data-frame were not returned by the original function). Saving the widget has been kept at the top-level of `plot_cluster_sina`.

Some tests were added for the function that converts the pldf object for presentation by ggplot2. No tests were added for the plot- or htmlwidget-construction phases.

Eventually, we will want to store the ggplot2 object as an .Rds, and may remove the construction / saving of the htmlwidget from the ggplot object (this could be done by tfpbrowser).

### CODE REVIEW

**Comment by russHyde** [comment]
This PR is now set to merge into `staging-wp3`. I'll direct all PRs for workpackage3 into that staging branch to mitigate against issues arising from upstream:master receiving commits.

File: R/create_widget.R [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#' @param   width_svg,height_svg    Width and height for the output SVG.
+#' @param   sizing_width    Sets the width option in \code{ggiraph::opts_sizing}.
+#'
+#' @return  A \code{girafe} / \code{htmlwidget} object for use in a web browser.
+
+create_widget <- function(ggobj,
+                          width_svg,
+                          height_svg,
+                          sizing_width = 0.8) {
+  tooltip_css <- paste0(
```
Comment (jr-nicola) [inline R/create_widget.R:13]: I'm happy to add this into {tfpscanner} for now, but this code will eventually be moved into {tfpbrowser} instead.

File: R/plot_cluster_sina.R [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+#' @param varx  String. Which variable in \code{pldf} should be plotted on the y-axis.
+#' @param mut_regexp  String. Regular-expression(s) for restricting the plot to a subset of the rows
+#'   in \code{pldf}. Only rows with an "allmuts" entry that matches one of these regular expressions
+#'   will be presented. If NULL, all rows are presented.
+#' @param lineage_regexp  String. Regular-expression(s) for restricting the lineages that are
+#'   presented in the plot. Only those rows of \code{pldf} with a "lineage" entry that matches one
+#'   of these regular expressions will be presented. If NULL, all rows are presented.
+
+plot_cluster_sina <- function(pldf,
+                              output_dir,
```
Comment (jr-nicola) [inline R/plot_cluster_sina.R:16]: Additional argument added `output_dir`. I've checked this is called and does the same thing. Looks good to me

File: R/plot_cluster_sina.R [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+  p1 <- create_cluster_sina_ggplot(sc1, y_lab = varx)
+
+  g1 <- create_widget(
+    ggobj = p1,
+    width_svg = 8,
+    height_svg = 8
+  )
+
+  htmlwidgets::saveWidget(g1,
```
Comment (jr-nicola) [inline R/plot_cluster_sina.R:35]: Happy to leave this in for now as well to make changes step-by-step - though this line will be unnecessary when we save the ggiraph objects as .rds files.

File: R/treeview.R [hunk 4]
```diff
@@ -562,21 +481,24 @@ treeview <- function(e0,
     suppressWarnings(.plot_tree(vn, mut_regex = mutations))
   }
 
-  suppressWarnings(
-    .pl_cluster_sina(pldf,
+  suppressWarnings({
+    plot_cluster_sina(
+      pldf,
+      output_dir = output_dir,
```
Comment (jr-nicola) [inline R/treeview.R:487]: Assuming this just uses the default value of `varx = "logistic_growth_rate"` hence why the argument isn't specified. Same as original code so fine with this :)

File: tests/testthat/test-plot_cluster_sina.R [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+  it("returns a '<mutation>_<lineage>' string", {
+    ml <- get_mutation_and_lineage(df, mut_regexp = "S:A222V", lineage_regexp = "AY\\.43")
+    expect_equal(
+      ml,
+      c("_AY\\.43", "S:A222V_AY\\.43", "_")
+    )
+  })
+  it("is just an underscore if neither regexp pattern is specified", {
+    ml <- get_mutation_and_lineage(df, mut_regexp = NULL, lineage_regexp = NULL)
```
Comment (jr-nicola) [inline tests/testthat/test-plot_cluster_sina.R:57]: Should there be an `expect_equal()` here?

**Review by jr-nicola — COMMENTED** [review]
Looks good to me overall - just one main query on one of the tests :)

File: tests/testthat/test-plot_cluster_sina.R (hunk 5 above)
Comment (russHyde) [reply tests/testthat/test-plot_cluster_sina.R:57]: Yes. Thanks

Comment (russHyde) [reply tests/testthat/test-plot_cluster_sina.R:57]: Fixed

**Comment by russHyde** [comment]
@jr-nicola I've addressed the test and have fixed the NEWS conflict. Could you approve please.

**Review by jr-nicola — APPROVED** [review]
Looks good to me

**Pull request merged**